zerver/lib/api.py

```diff
@@ -0,0 +1,334 @@
+            assert (
+                self.whence is None
+            ), "argument_type_is_body is incompatible with a non-default whence"
+
+
+# TypeAliases for common Annotated types
+
+JSON = Annotated[T, Metadata(convert_from_json=True)]
+WebhookPayload = Annotated[T, Metadata(convert_from_json=True, argument_type_is_body=True)]
+PathOnly = Annotated[T, Metadata(path_only=True)]
+
+# Implementation
+
+
+class _NotSpecified:
+    pass
+
+
+NotSpecified = _NotSpecified()
+
+
+@frozen
+class FuncParam(Generic[T]):
+    # Metadata retrieved from the Annotated type.
+    metadata: Metadata
+    # Name of the function parameter as defined in the original function.
+    param_name: str
+    # Inspected underlying type of the parameter unwarping the Annotated type if
+    # there is one.
+    param_type: Type[T]
+    # Name of the corresponding variable in the request data to look for. When
+    # argument_type_is_body is True, this is set to "request".
+    request_var_name: str
+    # Default value of the parameter.
+    default: Union[T, _NotSpecified] = NotSpecified
+
+
+@frozen
+class ViewFuncInfo:
+    view_func_full_name: str
+    parameters: Sequence[FuncParam[object]]
+    # Whether a request is expected to be processed by another function
+    # decorated as an endpoint or not.
+    no_nesting: bool
+
+
+@frozen
+class CallInfo:
+    kwargs_to_update: Mapping[str, object]
+    processed_request_var_names: Set[str]
+    errors: Sequence[str]
+
+
+IGNORED_PARAMS = frozenset(["request", "user_profile", "args", "kwargs"])
+
+
+def parse_view_func_signature(
+    view_func: Callable[Concatenate[HttpRequest, ParamT], object]
+) -> ViewFuncInfo:
+    """This is responsible for inspecting the function signature and getting the
+    metadata from the parameters. We want to keep this function as pure as
+    possible not leaking side effects to the global state. Side effects should
+    be emitted separately after the ViewFuncInfo is returned.
+    """
+    type_hints = get_type_hints(view_func, include_extras=True)
+    parameters = inspect.signature(view_func).parameters
+    view_func_full_name = ".".join([view_func.__module__, view_func.__name__])
+
+    process_parameters: List[FuncParam[object]] = []
+
+    for param_name, parameter in parameters.items():
+        assert param_name in type_hints
+        if param_name in IGNORED_PARAMS:
+            continue
+
+        param_default = parameter.default
+        # inspect._empty is the internal type used by inspect to indicate not
+        # specified defaults.
+        if param_default is inspect._empty:
+            param_default = NotSpecified
+
+        # defaulting a value to None automatically wraps the type annotation
+        # with Optional. We explicitly unwrap it for the case of Annotated,
+        # which otherwise causes undesired behaviors.
+        # This is fixed in Python 3.11: https://github.com/python/cpython/issues/90353
+        param_type = type_hints[param_name]
```

Comment:
My personal preference would be to extract this chunk of code into an inner helper function called `process_single_parameter` or something.  There are tradeoffs here, so this is a suggestion not a request.